tools/clang/lib/SPIRV/SpirvBuilder.cpp

```diff
@@ -1633,48 +1639,53 @@ std::vector<uint32_t> SpirvBuilder::takeModule() {
   addModuleInitCallToEntryPoints();
 
   // Run necessary visitor passes first
-  LiteralTypeVisitor literalTypeVisitor(astContext, context, spirvOptions);
-  LowerTypeVisitor lowerTypeVisitor(astContext, context, spirvOptions);
-  CapabilityVisitor capabilityVisitor(astContext, context, spirvOptions, *this,
-                                      featureManager);
-  RelaxedPrecisionVisitor relaxedPrecisionVisitor(context, spirvOptions);
-  PreciseVisitor preciseVisitor(context, spirvOptions);
-  NonUniformVisitor nonUniformVisitor(context, spirvOptions);
-  RemoveBufferBlockVisitor removeBufferBlockVisitor(
-      astContext, context, spirvOptions, featureManager);
-  EmitVisitor emitVisitor(astContext, context, spirvOptions, featureManager);
-
-  mod->invokeVisitor(&literalTypeVisitor, true);
+  if (astContext) {
+    LiteralTypeVisitor literalTypeVisitor(*astContext, context, spirvOptions);
+    mod->invokeVisitor(&literalTypeVisitor, true);
+  }
 
   // Propagate NonUniform decorations
+  NonUniformVisitor nonUniformVisitor(context, spirvOptions);
   mod->invokeVisitor(&nonUniformVisitor);
 
   // Lower types
-  mod->invokeVisitor(&lowerTypeVisitor);
-
-  // Generate debug types (if needed)
-  if (spirvOptions.debugInfoRich) {
-    DebugTypeVisitor debugTypeVisitor(astContext, context, spirvOptions, *this,
-                                      lowerTypeVisitor);
-    SortDebugInfoVisitor sortDebugInfoVisitor(context, spirvOptions);
-    mod->invokeVisitor(&debugTypeVisitor);
-    mod->invokeVisitor(&sortDebugInfoVisitor);
+  if (astContext) {
+    LowerTypeVisitor lowerTypeVisitor(*astContext, context, spirvOptions);
```

Comment:
I am worried that this can be confusing.

If we do not set `astContext` for dxil-to-spirv, just returning this method when `astContext == nullptr` (with an error message explaining that this method expects `astContext != nullptr`).

I'd create a separate function for `EmitVisitor` and other visitors for dxil-to-spirv.
e.g., `takeModuleForDxilToSpv()` ??